// ng-jhipster/esm5/pipe/filter.pipe.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 Copyright 2013-2019 the original author or authors from the JHipster project.

 This file is part of the JHipster project, see https://www.jhipster.tech/
 for more information.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
import { Pipe } from '@angular/core';
var JhiFilterPipe = /** @class */ (function () {
    function JhiFilterPipe() {
    }
    /**
     * @param {?} input
     * @param {?} filter
     * @param {?} field
     * @return {?}
     */
    JhiFilterPipe.prototype.transform = /**
     * @param {?} input
     * @param {?} filter
     * @param {?} field
     * @return {?}
     */
    function (input, filter, field) {
        if (!filter) {
            return input;
        }
        /** @type {?} */
        var type = typeof filter;
        if (type === 'string') {
            if (field) {
                return input.filter(this.filterByStringAndField(filter, field));
            }
            return input.filter(this.filterByString(filter));
        }
        if (type === 'object') {
            return input.filter(this.filterByObject(filter));
        }
    };
    /**
     * @private
     * @param {?} filter
     * @param {?} field
     * @return {?}
     */
    JhiFilterPipe.prototype.filterByStringAndField = /**
     * @private
     * @param {?} filter
     * @param {?} field
     * @return {?}
     */
    function (filter, field) {
        return function (value) {
            return !filter || (value[field] && value[field].toLowerCase().indexOf(filter.toLowerCase()) !== -1);
        };
    };
    // adapted from https://github.com/VadimDez/ng2-filter-pipe
    // adapted from https://github.com/VadimDez/ng2-filter-pipe
    /**
     * @private
     * @param {?} filter
     * @return {?}
     */
    JhiFilterPipe.prototype.filterByString = 
    // adapted from https://github.com/VadimDez/ng2-filter-pipe
    /**
     * @private
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        return function (value) {
            return !filter || value.toLowerCase().indexOf(filter.toLowerCase()) !== -1;
        };
    };
    /**
     * @private
     * @param {?} filter
     * @return {?}
     */
    JhiFilterPipe.prototype.filterDefault = /**
     * @private
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        return function (value) {
            return !filter || filter === value;
        };
    };
    /**
     * @private
     * @param {?} filter
     * @return {?}
     */
    JhiFilterPipe.prototype.filterByObject = /**
     * @private
     * @param {?} filter
     * @return {?}
     */
    function (filter) {
        var _this = this;
        return function (value) {
            var e_1, _a;
            /** @type {?} */
            var keys = Object.keys(filter);
            try {
                for (var keys_1 = tslib_1.__values(keys), keys_1_1 = keys_1.next(); !keys_1_1.done; keys_1_1 = keys_1.next()) {
                    var key = keys_1_1.value;
                    /** @type {?} */
                    var type = typeof value[key];
                    /** @type {?} */
                    var isMatching = void 0;
                    if (type === 'string') {
                        isMatching = _this.filterByString(filter[key])(value[key]);
                    }
                    else if (type === 'object') {
                        isMatching = _this.filterByObject(filter[key])(value[key]);
                    }
                    else {
                        isMatching = _this.filterDefault(filter[key])(value[key]);
                    }
                    if (!isMatching) {
                        return false;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (keys_1_1 && !keys_1_1.done && (_a = keys_1.return)) _a.call(keys_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return true;
        };
    };
    JhiFilterPipe.decorators = [
        { type: Pipe, args: [{ name: 'filter', pure: false },] }
    ];
    return JhiFilterPipe;
}());
export { JhiFilterPipe };
