// ng-jhipster/esm2015/service/alert.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 Copyright 2013-2019 the original author or authors from the JHipster project.

 This file is part of the JHipster project, see https://www.jhipster.tech/
 for more information.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
import { Injectable, Sanitizer, SecurityContext } from '@angular/core';
import { TranslateService } from '@ngx-translate/core';
import { JhiConfigService } from '../config.service';
import * as i0 from "@angular/core";
import * as i1 from "../config.service";
import * as i2 from "@ngx-translate/core";
/**
 * @record
 */
export function JhiAlert() { }
if (false) {
    /** @type {?|undefined} */
    JhiAlert.prototype.id;
    /** @type {?} */
    JhiAlert.prototype.type;
    /** @type {?} */
    JhiAlert.prototype.msg;
    /** @type {?|undefined} */
    JhiAlert.prototype.params;
    /** @type {?|undefined} */
    JhiAlert.prototype.timeout;
    /** @type {?|undefined} */
    JhiAlert.prototype.toast;
    /** @type {?|undefined} */
    JhiAlert.prototype.position;
    /** @type {?|undefined} */
    JhiAlert.prototype.scoped;
    /** @type {?|undefined} */
    JhiAlert.prototype.close;
}
export class JhiAlertService {
    /**
     * @param {?} sanitizer
     * @param {?} configService
     * @param {?} translateService
     */
    constructor(sanitizer, configService, translateService) {
        this.sanitizer = sanitizer;
        this.configService = configService;
        this.translateService = translateService;
        /** @type {?} */
        const config = this.configService.getConfig();
        this.toast = config.alertAsToast;
        this.i18nEnabled = config.i18nEnabled;
        this.alertId = 0; // unique id for each alert. Starts from 0.
        this.alerts = [];
        this.timeout = config.alertTimeout;
    }
    /**
     * @return {?}
     */
    clear() {
        this.alerts.splice(0, this.alerts.length);
    }
    /**
     * @return {?}
     */
    get() {
        return this.alerts;
    }
    /**
     * @param {?} msg
     * @param {?=} params
     * @param {?=} position
     * @return {?}
     */
    success(msg, params, position) {
        return this.addAlert({
            type: 'success',
            msg,
            params,
            timeout: this.timeout,
            toast: this.isToast(),
            position
        }, []);
    }
    /**
     * @param {?} msg
     * @param {?=} params
     * @param {?=} position
     * @return {?}
     */
    error(msg, params, position) {
        return this.addAlert({
            type: 'danger',
            msg,
            params,
            timeout: this.timeout,
            toast: this.isToast(),
            position
        }, []);
    }
    /**
     * @param {?} msg
     * @param {?=} params
     * @param {?=} position
     * @return {?}
     */
    warning(msg, params, position) {
        return this.addAlert({
            type: 'warning',
            msg,
            params,
            timeout: this.timeout,
            toast: this.isToast(),
            position
        }, []);
    }
    /**
     * @param {?} msg
     * @param {?=} params
     * @param {?=} position
     * @return {?}
     */
    info(msg, params, position) {
        return this.addAlert({
            type: 'info',
            msg,
            params,
            timeout: this.timeout,
            toast: this.isToast(),
            position
        }, []);
    }
    /**
     * @param {?} alertOptions
     * @param {?} extAlerts
     * @return {?}
     */
    addAlert(alertOptions, extAlerts) {
        alertOptions.id = this.alertId++;
        if (this.i18nEnabled && alertOptions.msg) {
            alertOptions.msg = this.translateService.instant(alertOptions.msg, alertOptions.params);
        }
        /** @type {?} */
        const alert = this.factory(alertOptions);
        if (alertOptions.timeout && alertOptions.timeout > 0) {
            setTimeout(() => {
                this.closeAlert(alertOptions.id, extAlerts);
            }, alertOptions.timeout);
        }
        return alert;
    }
    /**
     * @param {?} id
     * @param {?=} extAlerts
     * @return {?}
     */
    closeAlert(id, extAlerts) {
        /** @type {?} */
        const thisAlerts = extAlerts && extAlerts.length > 0 ? extAlerts : this.alerts;
        return this.closeAlertByIndex(thisAlerts.map(e => e.id).indexOf(id), thisAlerts);
    }
    /**
     * @param {?} index
     * @param {?} thisAlerts
     * @return {?}
     */
    closeAlertByIndex(index, thisAlerts) {
        return thisAlerts.splice(index, 1);
    }
    /**
     * @return {?}
     */
    isToast() {
        return this.toast;
    }
    /**
     * @private
     * @param {?} alertOptions
     * @return {?}
     */
    factory(alertOptions) {
        /** @type {?} */
        const alert = {
            type: alertOptions.type,
            msg: this.sanitizer.sanitize(SecurityContext.HTML, alertOptions.msg),
            id: alertOptions.id,
            timeout: alertOptions.timeout,
            toast: alertOptions.toast,
            position: alertOptions.position ? alertOptions.position : 'top right',
            scoped: alertOptions.scoped,
            close: (alerts) => {
                return this.closeAlert(alertOptions.id, alerts);
            }
        };
        if (!alert.scoped) {
            this.alerts.push(alert);
        }
        return alert;
    }
}
JhiAlertService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
JhiAlertService.ctorParameters = () => [
    { type: Sanitizer },
    { type: JhiConfigService },
    { type: TranslateService }
];
/** @nocollapse */ JhiAlertService.ngInjectableDef = i0.defineInjectable({ factory: function JhiAlertService_Factory() { return new JhiAlertService(i0.inject(i0.Sanitizer), i0.inject(i1.JhiConfigService), i0.inject(i2.TranslateService)); }, token: JhiAlertService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    JhiAlertService.prototype.alertId;
    /**
     * @type {?}
     * @private
     */
    JhiAlertService.prototype.alerts;
    /**
     * @type {?}
     * @private
     */
    JhiAlertService.prototype.timeout;
    /**
     * @type {?}
     * @private
     */
    JhiAlertService.prototype.toast;
    /**
     * @type {?}
     * @private
     */
    JhiAlertService.prototype.i18nEnabled;
    /**
     * @type {?}
     * @private
     */
    JhiAlertService.prototype.sanitizer;
    /**
     * @type {?}
     * @private
     */
    JhiAlertService.prototype.configService;
    /**
     * @type {?}
     * @private
     */
    JhiAlertService.prototype.translateService;
}
