// ng-jhipster/esm2015/service/event-manager.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 Copyright 2013-2019 the original author or authors from the JHipster project.

 This file is part of the JHipster project, see https://www.jhipster.tech/
 for more information.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { filter, share } from 'rxjs/operators';
import * as i0 from "@angular/core";
/**
 * An utility class to manage RX events
 */
export class JhiEventManager {
    constructor() {
        this.observable = Observable.create((observer) => {
            this.observer = observer;
        }).pipe(share());
    }
    /**
     * Method to broadcast the event to observer
     * @param {?} event
     * @return {?}
     */
    broadcast(event) {
        if (this.observer != null) {
            this.observer.next(event);
        }
    }
    /**
     * Method to subscribe to an event with callback
     * @param {?} eventName
     * @param {?} callback
     * @return {?}
     */
    subscribe(eventName, callback) {
        /** @type {?} */
        const subscriber = this.observable
            .pipe(filter(event => {
            return event.name === eventName;
        }))
            .subscribe(callback);
        return subscriber;
    }
    /**
     * Method to unsubscribe the subscription
     * @param {?} subscriber
     * @return {?}
     */
    destroy(subscriber) {
        subscriber.unsubscribe();
    }
}
JhiEventManager.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
JhiEventManager.ctorParameters = () => [];
/** @nocollapse */ JhiEventManager.ngInjectableDef = i0.defineInjectable({ factory: function JhiEventManager_Factory() { return new JhiEventManager(); }, token: JhiEventManager, providedIn: "root" });
if (false) {
    /** @type {?} */
    JhiEventManager.prototype.observable;
    /** @type {?} */
    JhiEventManager.prototype.observer;
}
